models/inventory/StockManager.ts

```diff
@@ -132,16 +132,54 @@ export class StockManager {
     }
 
     const date = details.date.toISOString();
+    const formattedDate = this.fyo.format(details.date, 'Datetime');
+    const isItemHasBatchNumber = await this.fyo.getValue(
+      'Item',
+      details.item,
+      'hasBatchNumber'
+    );
+
+    if (isItemHasBatchNumber && !this.isCancelled) {
```

Comment:
Don't write nested ifs, readability takes a hit. If you have to nest an if, the body of the if block should be a function call.

(p.d. don't push changes I'll fix them in a separate PR, this one has been open for too long)